packages/drivers/odsp-driver/src/test/epochTestsWithRedemption.spec.ts

```diff
@@ -0,0 +1,83 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { TelemetryNullLogger } from "@fluidframework/common-utils";
+import { OdspErrorType } from "@fluidframework/odsp-doclib-utils";
+import { IOdspResolvedUrl } from "../contracts";
+import { EpochTrackerWithRedemption } from "../epochTracker";
+import { ICacheEntry, LocalPersistentCache, LocalPersistentCacheAdapter } from "../odspCache";
+import { getHashedDocumentId } from "../odspUtils";
+import { mockFetch, notFound } from "./mockFetch";
+
+describe("Tests for Epoch Tracker With Redemption", () => {
+    const siteUrl = "https://microsoft.sharepoint-df.com/siteUrl";
+    const driveId = "driveId";
+    const itemId = "fileId";
+    let epochTracker: EpochTrackerWithRedemption;
+    let cache: LocalPersistentCacheAdapter;
+    const hashedDocumentId = getHashedDocumentId(driveId, itemId);
+    beforeEach(() => {
+        cache = new LocalPersistentCacheAdapter(new LocalPersistentCache());
+        epochTracker = new EpochTrackerWithRedemption(cache, new TelemetryNullLogger());
+        const resolvedUrl = ({ siteUrl, driveId, itemId } as any) as IOdspResolvedUrl;
+        epochTracker.fileEntry = {
+            docId: hashedDocumentId,
+            resolvedUrl,
+        };
+    });
+
+    it("joinSession call should succeed on retrying after any network call to the file succeeds", async () => {
+        let success: boolean = true;
+        const resolvedUrl = ({ siteUrl, driveId, itemId } as any) as IOdspResolvedUrl;
+        const cacheEntry1: ICacheEntry = {
+            key:"key1",
+            type: "snapshot",
+            file: { docId: hashedDocumentId, resolvedUrl } };
+        cache.put(cacheEntry1, { value: "val1", fluidEpoch: "epoch1", version: "0.1" }, 0);
+
+        try {
+            // We will trigger a successful call to return the value set in the cache after the failed joinSession call
+            // eslint-disable-next-line @typescript-eslint/no-misused-promises
+            setTimeout(async () => mockFetch({}, async () => {
```

Comment:
Right now I'm using setTimeout to launch a second request after joinSession fails since the joinSession call only returns after the retry has finished. Would you know of a better way to do this? I was trying to also do Promise.all to launch them concurrently but then there is a race to mock the fetch function since both the 200 & the 404 will try to overwrite each others mocks

Reply:
I do not see good way to solve it without exposing more state out of EpochTrackerWithRedemption.
You could drive a test-only class from it and expose more data for test to use? Or even simply patch it on the fly (it's Javascript!) I do not see an easy way other than overwrite this.treesLatestDeferral with some custom class that would expose a signal when someone started to wait on promise, and that signal would be used to launch another request here.
It's doable, just a bit ugly and depends on implementation details of the class.

Reply:
Added in custom class to use for treesLatestDeferral